From 9bb842fa925546a9d676a9001a6eb98797aba307 Mon Sep 17 00:00:00 2001
From: William Wu <william.wu@rock-chips.com>
Date: Sat, 9 May 2020 15:38:16 +0800
Subject: [PATCH] usb: gadget: uvc: configfs: fix the desc of framebased format

Refer to the spec "USB_Video_Payload_Frame_Based_1.1", we
set the bBitsPerPixel of framebased format descriptor to 16,
it means that 16 bits per pixel used to specify color in the
decoded video frame. And set bVariableSize of framebased format
descriptor to 1, it means that the data within the frame is of
variable length from frame to frame is variable.

With this patch, the uvc H264 format can work normally on
Linux OS platform.

Signed-off-by: William Wu <william.wu@rock-chips.com>
Change-Id: Ibc25b5f67efb8cfcae9d66a509399cfe896f7943
---
 drivers/usb/gadget/function/uvc_configfs.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/drivers/usb/gadget/function/uvc_configfs.c b/drivers/usb/gadget/function/uvc_configfs.c
index 1a529b0f7a9b..058a3fbbd220 100644
--- a/drivers/usb/gadget/function/uvc_configfs.c
+++ b/drivers/usb/gadget/function/uvc_configfs.c
@@ -2128,12 +2128,13 @@ static struct config_group *uvcg_framebased_make(struct config_group *group,
 	f->desc.bDescriptorType		= USB_DT_CS_INTERFACE;
 	f->desc.bDescriptorSubType	= UVC_VS_FORMAT_FRAME_BASED;
 	memcpy(f->desc.guidFormat, guid, sizeof(guid));
-	f->desc.bBitsPerPixel		= 0;
+	f->desc.bBitsPerPixel		= 16;
 	f->desc.bDefaultFrameIndex	= 1;
 	f->desc.bAspectRatioX		= 0;
 	f->desc.bAspectRatioY		= 0;
 	f->desc.bmInterfaceFlags	= 0;
 	f->desc.bCopyProtect		= 0;
+	f->desc.bVariableSize		= 1;
 
 	f->fmt.type = UVCG_FRAMEBASED;
 	config_group_init_type_name(&f->fmt.group, name,
-- 
2.35.3

